substrate/frame/transaction-payment/src/lib.rs

```diff
@@ -825,68 +849,130 @@ impl<T: Config> core::fmt::Debug for ChargeTransactionPayment<T> {
 	}
 }
 
-impl<T: Config> SignedExtension for ChargeTransactionPayment<T>
+/// The info passed between the validate and prepare steps for the `ChargeAssetTxPayment` extension.
+#[derive(RuntimeDebugNoBound)]
+pub enum Val<T: Config> {
+	Charge {
+		tip: BalanceOf<T>,
+		// who paid the fee
+		who: T::AccountId,
+		// transaction fee
+		fee: BalanceOf<T>,
+	},
+	NoCharge,
+}
+
+/// The info passed between the prepare and post-dispatch steps for the `ChargeAssetTxPayment`
+/// extension.
+pub enum Pre<T: Config> {
+	Charge {
+		tip: BalanceOf<T>,
+		// who paid the fee
+		who: T::AccountId,
+		// imbalance resulting from withdrawing the fee
+		imbalance: <<T as Config>::OnChargeTransaction as OnChargeTransaction<T>>::LiquidityInfo,
+	},
+	NoCharge {
+		// weight initially estimated by the extension, to be refunded
+		refund: Weight,
+	},
```

Comment:
Both could just be an `Option` that with `None` refunds the weight.

Reply:
This one could be an option because we know there is only one weight this extension can have, but for example in the case of both extensions in `pallet-asset-tx-payment` and `pallet-asset-conversion-tx-payment` this wouldn't work because there would be no way for you to figure out what weight you used before dispatch if you don't provide it in `Pre`.

Anyway this is only used internally in the extension code and is just a bit more explicit while being consistent with other extensions, I'd keep as is.